### ISSUE TITLE
kv: query repeatedly failing with error "truncation resulted in empty batch"

### ISSUE THREAD

**Issue by a-robinson** [cockroachdb/cockroach#18174]
On blue-0001:

```sql
root@:26257/> SELECT * FROM system.rangelog where "rangeID" = 1 OR "otherRangeID" = 1 ORDER BY timestamp DESC;
pq: kv/dist_sender.go:915: truncation resulted in empty batch on /Table/13/1/2017-08-{29T05:43:22.823574Z/274972681302671366-30T13:49:10.277057Z/275351307422334986}: [txn: 49b3e27c], ReverseScan [/Table/13/1/2017-08-30T13:49:10.277057Z/275351307422334986,/Table/13/1/2017-08-30T14:36:33.205243Z/275360622898970627/6/1/NULL), ReverseScan [/Table/13/1/2017-08-23T11:16:32.574447Z/273339491307421696,/Table/13/1/2017-08-25T06:33:53.920764Z/273850151531053057)
```

A couple hours later, the same problem still persists every time I run the query:

```sql
root@:26257/> SELECT * FROM system.rangelog where "rangeID" = 1 OR "otherRangeID" = 1 ORDER BY timestamp DESC;
pq: kv/dist_sender.go:915: truncation resulted in empty batch on /Table/13/1/2017-08-{29T05:43:22.823574Z/274972681302671366-30T13:49:10.277057Z/275351307422334986}: [txn: 2d1acbe4], ReverseScan [/Table/13/1/2017-08-30T13:49:10.277057Z/275351307422334986,/Table/13/1/2017-08-30T14:37:02.46295Z/275360718781775878/6/1/NULL), ReverseScan [/Table/13/1/2017-08-23T11:16:32.574447Z/273339491307421696,/Table/13/1/2017-08-25T06:33:53.920764Z/273850151531053057)
root@:26257/> SELECT * FROM system.rangelog where "rangeID" = 1 OR "otherRangeID" = 1 ORDER BY timestamp DESC;
pq: kv/dist_sender.go:915: truncation resulted in empty batch on /Table/13/1/2017-08-{29T05:43:22.823574Z/274972681302671366-30T13:49:10.277057Z/275351307422334986}: [txn: a0b2a903], ReverseScan [/Table/13/1/2017-08-30T13:49:10.277057Z/275351307422334986,/Table/13/1/2017-08-30T16:06:44.953574Z/275378356147486726/3/1/NULL), ReverseScan [/Table/13/1/2017-08-26T15:27:29.330384Z/274238175159386115,/Table/13/1/2017-08-29T05:43:22.823574Z/274972681302671366)
```

Putting a limit on it or removing the ORDER BY fixes things (it's really slow, as covered in #18159, but it succeeds):

```sql
SELECT * FROM system.rangelog where "rangeID" = 1 OR "otherRangeID" = 1 ORDER BY timestamp DESC LIMIT 100;
[...]
(100 rows)

Time: 2m14.285688078s

root@:26257/> SELECT * FROM system.rangelog where "rangeID" = 1 OR "otherRangeID" = 1;
[...]
(151 rows)

Time: 5.350987635s
```

It's possible that 30f4050c1ebe3a6251cb9bef871417f30a1ad3f7 did something to unveil this.

@tschottdorf @nvanbenschoten

**Comment by a-robinson** [comment]
`EXPLAIN` output for the different commands:

```
root@:26257/> EXPLAIN SELECT * FROM system.rangelog where "rangeID" = 1 OR "otherRangeID" = 1 ORDER BY timestamp DESC;
+-------+---------+-------+------------------+
| Level |  Type   | Field |   Description    |
+-------+---------+-------+------------------+
|     0 | revscan |       |                  |
|     0 |         | table | rangelog@primary |
|     0 |         | spans | ALL              |
+-------+---------+-------+------------------+
root@:26257/> EXPLAIN SELECT * FROM system.rangelog where "rangeID" = 1 OR "otherRangeID" = 1 ORDER BY timestamp DESC LIMIT 100;
+-------+---------+-------+------------------+
| Level |  Type   | Field |   Description    |
+-------+---------+-------+------------------+
|     0 | limit   |       |                  |
|     1 | revscan |       |                  |
|     1 |         | table | rangelog@primary |
|     1 |         | spans | ALL              |
+-------+---------+-------+------------------+
root@:26257/> EXPLAIN SELECT * FROM system.rangelog where "rangeID" = 1 OR "otherRangeID" = 1;
+-------+------+-------+------------------+
| Level | Type | Field |   Description    |
+-------+------+-------+------------------+
|     0 | scan |       |                  |
|     0 |      | table | rangelog@primary |
|     0 |      | spans | ALL              |
+-------+------+-------+------------------+
```

**Comment by a-robinson** [comment]
Sorry for leaving out the `ORDER BY ASC` case. It works just fine, so seemingly something is up with the revscan.

```sql
SELECT * FROM system.rangelog where "rangeID" = 1 OR "otherRangeID" = 1 ORDER BY timestamp ASC;
[...]
(151 rows)

Time: 6.265527528s
root@:26257/> EXPLAIN SELECT * FROM system.rangelog where "rangeID" = 1 OR "otherRangeID" = 1 ORDER BY timestamp ASC;
+-------+------+-------+------------------+
| Level | Type | Field |   Description    |
+-------+------+-------+------------------+
|     0 | scan |       |                  |
|     0 |      | table | rangelog@primary |
|     0 |      | spans | ALL              |
+-------+------+-------+------------------+
```

**Comment by a-robinson** [comment]
I don't see anything obviously wrong in the session trace:

[trace-kv-bug.txt](https://github.com/cockroachdb/cockroach/files/1271725/trace-kv-bug.txt)

**Comment by tschottdorf** [comment]
Looking.

**Comment by tschottdorf** [comment]
Maybe I'm missing something obvious, but here's my half-baked theory (that I'm hastily writing down since I have to leave soon).

What sparked my curiosity is that the situation here looks one-offy, namely the error is essentially saying that `DistSender` was trying to restrict a request of the form `RevScan [b,c)` to a range `[a,b)` and naturally that didn't intersect.

How did we get there? Well, the batch has some more scans (at lower key ranges) and we're going reverse, so let's say the batch is really `RevScan[b,c) RevScan[a,aa)` and we've just handled basically the whole `[b,c)` using descriptor `[b,y)`. We should then hop to a descriptor that (at least) touches `aa`, but I *think* we hop into a gap.

To determine the new lower key to search a descriptor for, the following code runs:

```
seekKey, err = prev(ba /* [a,aa) + [b,c) */, ri.Desc().StartKey /* b */)
nextRS.EndKey = seekKey
```

I *think* you would want `seekKey` to be `aa` here, but I don't think that's the case:

```go
func TestFoo(t *testing.T) {
	var ba roachpb.BatchRequest
	args := &roachpb.ScanRequest{}
	args.Key, args.EndKey = roachpb.Key("b"), roachpb.Key("c")
	ba.Add(args)
	k, err := prev(ba, roachpb.RKey("b"))
	if err != nil {
		t.Fatal(err)
	}
	// --- FAIL: TestFoo (0.00s)
	//	batch_test.go:35: b
	t.Error(string(k))
}
```

So we're seeking at `b` instead of `aa`. The only checks to clear are those two:

```go
//                                                    [a,c)                 b
if !ri.NeedAnother(rs) || !nextRS.Key.Less(nextRS.EndKey) {
	return
}
```

and `NeedAnother` boils down to `rs.Key.Less(ri.desc.StartKey)` which is `a.Less(b)` which is also true.

Correspondingly the next descriptor looked up will touch `b`, and if we're unlucky then it won't touch `[a,aa)` and so we end up truncating a batch to a range it doesn't intersect.

If someone could sanity check the above, I'd appreciate it. Given confirmation, I'll start working on test+fix.

**Comment by a-robinson** [comment]
Yeah. This is my first time taking a close look at this code, but your explanation looks right to me.

**Comment by nvanbenschoten** [comment]
@tschottdorf that theory makes sense to me as well. Do we have a similar issue with forward `ScanRequests` when we call `seekKey, err = next(ba, ri.Desc().EndKey)`?

**Comment by tschottdorf** [comment]
Just a heads up that I've repro'ed this in a unit test:

```go
--- FAIL: TestMultiRangeGapReverse (0.04s)
	dist_sender_test.go:1271: kv/dist_sender.go:906: truncation resulted in empty batch on {b-c}: [txn: ed239b6b], ReverseScan ["a","b"), ReverseScan ["c","d")
```

I'll work on a fix and also verify the corresponding situation in forward scans.

**Comment by tschottdorf** [comment]
Fixed in #18351. I just re-ran the above query and it worked.
